[Security Solution] consolidate attack discovery + defend insights

joeypoon opened the pull request:
## Summary

Consolidates the common LangGraph output chunking logic used by Attack Discovery and Defend Insights. Functionality should be unchanged.

Applicable nodes, edges, and helpers were moved from Attack Discovery and Defend Insights to `x-pack/solutions/security/plugins/elastic_assistant/server/lib/langchain/output_chunking/` then deleted from the corresponding Attack Discovery / Defend Insight directories.

The retriever nodes were not moved since these had enough differences to make it difficult to consolidate and given that Defend Insight will be supporting additional data sources in the near future, I think it makes sense to keep this separate.

In order to consolidate the `GraphState`, some of the properties were renamed to be more generic. There is now a `BaseGraphState` which accepts a generic type that defines the insight type (`AttackDiscovery` or `DefendInsight`).


### Checklist

- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios

joeypoon committed:
[Security Solution] consolidate attack discovery + defend insights

Consolidates the common langgraph logic used by Attack Discovery and Defend Insights

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/289364)
* Commit: 82d6b079b51540f0d3a0e74f26be513ea3b550a7

### Failed CI Steps
* [Build Kibana Distribution](https://buildkite.com/elastic/kibana-pull-request/builds/289364#0195f5bd-5334-4928-8e8a-7456a6860960)

### History
* :broken_heart: [Build #289348](https://buildkite.com/elastic/kibana-pull-request/builds/289348) failed 5dae5762b05b0b82e89aa810ce16973227c01dd3

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/289364","number":289364,"commit":"82d6b079b51540f0d3a0e74f26be513ea3b550a7"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/289348","number":289348,"commit":"5dae5762b05b0b82e89aa810ce16973227c01dd3"}],"number":289364}
buildkite-pr-comment-->

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/289496)
* Commit: 91da06e5cf07b14e73946ca03eead290ba109b7f
* Kibana Serverless Image: `docker.elastic.co/kibana-ci/kibana-serverless:pr-216757-91da06e5cf07`

### History
* :broken_heart: [Build #289364](https://buildkite.com/elastic/kibana-pull-request/builds/289364) failed 82d6b079b51540f0d3a0e74f26be513ea3b550a7
* :broken_heart: [Build #289348](https://buildkite.com/elastic/kibana-pull-request/builds/289348) failed 5dae5762b05b0b82e89aa810ce16973227c01dd3

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"running","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/289496","number":289496,"commit":"91da06e5cf07b14e73946ca03eead290ba109b7f"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/289364","number":289364,"commit":"82d6b079b51540f0d3a0e74f26be513ea3b550a7"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/289348","number":289348,"commit":"5dae5762b05b0b82e89aa810ce16973227c01dd3"}],"number":289496}
buildkite-pr-comment-->

elasticmachine commented:
Pinging @elastic/security-defend-workflows (Team:Defend Workflows)

szwarckonrad reviewed (approved):
I went through the changes and they look good to me - I didn’t spot any red flags, though my understanding is limited. I left a few comments, mostly just to flag specific sections for you to double-check.

elasticmachine commented:
merge conflict between base and head

szwarckonrad committed:
Merge branch 'main' into chore/langgraph-consolidation

# Conflicts:
#	x-pack/solutions/security/plugins/elastic_assistant/server/routes/defend_insights/helpers.ts

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/296996)
* Commit: f734bcd06dc7465d059f2a0fcf4e020e7335d258
* Kibana Serverless Image: `docker.elastic.co/kibana-ci/kibana-serverless:pr-216757-f734bcd06dc7`

### History
* :green_heart: [Build #289496](https://buildkite.com/elastic/kibana-pull-request/builds/289496) succeeded 91da06e5cf07b14e73946ca03eead290ba109b7f
* :broken_heart: [Build #289364](https://buildkite.com/elastic/kibana-pull-request/builds/289364) failed 82d6b079b51540f0d3a0e74f26be513ea3b550a7
* :broken_heart: [Build #289348](https://buildkite.com/elastic/kibana-pull-request/builds/289348) failed 5dae5762b05b0b82e89aa810ce16973227c01dd3

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"running","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/296996","number":296996,"commit":"f734bcd06dc7465d059f2a0fcf4e020e7335d258"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/289496","number":289496,"commit":"91da06e5cf07b14e73946ca03eead290ba109b7f"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/289364","number":289364,"commit":"82d6b079b51540f0d3a0e74f26be513ea3b550a7"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/289348","number":289348,"commit":"5dae5762b05b0b82e89aa810ce16973227c01dd3"}],"number":296996}
buildkite-pr-comment-->

andrew-goldstein commented on x-pack/solutions/security/plugins/elastic_assistant/server/lib/langchain/graphs/index.ts:
```diff
@@ -68,3 +71,37 @@ export const ASSISTANT_GRAPH_MAP: Record<string, GraphMetadata> = {
+export type GraphInsightTypes = AttackDiscovery | DefendInsight;
+
+export interface BaseGraphState<T extends GraphInsightTypes> {
+  insights: T[] | null;
+  prompt: string;
+  anonymizedDocuments: Document[];
+  combinedGenerations: string;
+  combinedRefinements: string;
+  errors: string[];
+  generationAttempts: number;
+  generations: string[];
+  hallucinationFailures: number;
+  maxGenerationAttempts: number;
+  maxHallucinationFailures: number;
+  maxRepeatedGenerations: number;
+  refinements: string[];
+  refinePrompt: string;
+  continuePrompt: string;
+  replacements: Replacements;
+  unrefinedResults: T[] | null;
+  filter?: Record<string, unknown> | null;
+  start?: DateMath;
+  end?: DateMath;
```
nit: sort alphabetically for readability:

```typescript
  anonymizedDocuments: Document[];
  combinedGenerations: string;
  combinedRefinements: string;
  continuePrompt: string;
  end?: DateMath;
  errors: string[];
  filter?: Record<string, unknown> | null;
  generationAttempts: number;
  generations: string[];
  hallucinationFailures: number;
  insights: T[] | null;
  maxGenerationAttempts: number;
  maxHallucinationFailures: number;
  maxRepeatedGenerations: number;
  prompt: string;
  refinements: string[];
  refinePrompt: string;
  replacements: Replacements;
  start?: DateMath;
  unrefinedResults: T[] | null;
```

andrew-goldstein committed:
- fixed a circular dependency on `NodeType` that caused evaluations to fail
- included a missing fix from https://github.com/elastic/kibana/pull/215816
- fixed an issue where Attack discovery evaluations returned lower-than-expected scores due to renamed graph input overrides

prodsecmachine commented:
![](https://res.cloudinary.com/snyk/image/upload/r-d/scm-platform/snyk-pull-requests/pr-banner-default.svg)

### :tada: **Snyk checks have passed. No issues have been found so far.**

:white_check_mark: **security/snyk** check is complete. No issues have been found. [(View Details)](https://app.snyk.io/org/kibana/pr-checks/9f7d753f-71e3-4da8-a4fc-6725d202425f?refs=48ded9315a027bc5a7eed2a1eb07741cb976af98&source=prChecksComment)

:white_check_mark: **license/snyk** check is complete. No issues have been found. [(View Details)](https://app.snyk.io/org/kibana/pr-checks/9f7d753f-71e3-4da8-a4fc-6725d202425f/license?refs=48ded9315a027bc5a7eed2a1eb07741cb976af98&source=prChecksComment)

andrew-goldstein commented on x-pack/solutions/security/plugins/elastic_assistant/server/lib/langchain/output_chunking/nodes/refine/helpers/get_combined_refine_prompt/index.ts:
```diff
@@ -5,31 +5,36 @@
 
 """
-${JSON.stringify(unrefinedResults, null, 2)}
+\`\`\`json
+{
+  "insights": ${JSON.stringify(unrefinedResults, null, 2)}
+}
+\`\`\`
```
This fixes an intermittent refinement step error, detailed in https://github.com/elastic/kibana/pull/215816 and addressed in commit https://github.com/elastic/kibana/pull/216757/commits/48ded9315a027bc5a7eed2a1eb07741cb976af98

andrew-goldstein commented on x-pack/solutions/security/plugins/elastic_assistant/server/lib/langchain/graphs/constants.ts:
```diff
@@ -0,0 +1,12 @@
+ * 2.0.
+ */
+
+export const NodeType = {
+  GENERATE_NODE: 'generate',
+  REFINE_NODE: 'refine',
+  RETRIEVE_ANONYMIZED_DOCS_NODE: 'retrieve_anonymized_docs',
+} as const;
```
`NodeType` moved here from `x-pack/solutions/security/plugins/elastic_assistant/server/lib/langchain/graphs/index.ts` because that file imports `getDefaultAttackDiscoveryGraph`, and in 
In `default_attack_discovery_graph/index.ts`, it was importing `NodeType` from `../../../langchain/graphs`, which created a circular depdency, resulting in a runtime error when running AD evals.

andrew-goldstein replied:
(via commit https://github.com/elastic/kibana/pull/216757/commits/48ded9315a027bc5a7eed2a1eb07741cb976af98)

andrew-goldstein commented on x-pack/solutions/security/plugins/elastic_assistant/server/lib/attack_discovery/evaluation/helpers/get_graph_input_overrides/index.ts:
```diff
@@ -5,25 +5,27 @@
+
+  // Fallback to and rename the root level legacy properties `anonymizedAlerts`,
+  // and `attackDiscoveries` to the new graph property names:
+  const anonymizedDocuments = validatedInput.anonymizedDocuments ?? validatedInput.anonymizedAlerts;
+  const insights = validatedInput.insights ?? validatedInput.attackDiscoveries;
```
This fixes an issue where Attack discovery evaluations returned lower-than-expected scores because the legacy Langsmith dataset examples were not overriding the new graph state properties. Addressed in https://github.com/elastic/kibana/pull/216757/commits/48ded9315a027bc5a7eed2a1eb07741cb976af98

andrew-goldstein reviewed (approved):
Thanks @joeypoon for consolidating the Attack discovery + Defend insights output chunking! 🙏 They started to diverge (where one didn't include fixes from the other), so your effort is already making an impact.
✅ Desk tested locally
LGTM 🚀

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/298027)
* Commit: 48ded9315a027bc5a7eed2a1eb07741cb976af98
* Kibana Serverless Image: `docker.elastic.co/kibana-ci/kibana-serverless:pr-216757-48ded9315a02`

### History
* :yellow_heart: [Build #296996](https://buildkite.com/elastic/kibana-pull-request/builds/296996) was flaky f734bcd06dc7465d059f2a0fcf4e020e7335d258
* :green_heart: [Build #289496](https://buildkite.com/elastic/kibana-pull-request/builds/289496) succeeded 91da06e5cf07b14e73946ca03eead290ba109b7f
* :broken_heart: [Build #289364](https://buildkite.com/elastic/kibana-pull-request/builds/289364) failed 82d6b079b51540f0d3a0e74f26be513ea3b550a7
* :broken_heart: [Build #289348](https://buildkite.com/elastic/kibana-pull-request/builds/289348) failed 5dae5762b05b0b82e89aa810ce16973227c01dd3

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"running","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/298027","number":298027,"commit":"48ded9315a027bc5a7eed2a1eb07741cb976af98"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/296996","number":296996,"commit":"f734bcd06dc7465d059f2a0fcf4e020e7335d258"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/289496","number":289496,"commit":"91da06e5cf07b14e73946ca03eead290ba109b7f"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/289364","number":289364,"commit":"82d6b079b51540f0d3a0e74f26be513ea3b550a7"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/289348","number":289348,"commit":"5dae5762b05b0b82e89aa810ce16973227c01dd3"}],"number":298027}
buildkite-pr-comment-->

joeypoon committed:
Merge branch 'main' into chore/langgraph-consolidation

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/300345)
* Commit: ee72370a5a87d28338702cfbe4f1506cb2ff982f
* Kibana Serverless Image: `docker.elastic.co/kibana-ci/kibana-serverless:pr-216757-ee72370a5a87`

### History
* :green_heart: [Build #298027](https://buildkite.com/elastic/kibana-pull-request/builds/298027) succeeded 48ded9315a027bc5a7eed2a1eb07741cb976af98
* :yellow_heart: [Build #296996](https://buildkite.com/elastic/kibana-pull-request/builds/296996) was flaky f734bcd06dc7465d059f2a0fcf4e020e7335d258
* :green_heart: [Build #289496](https://buildkite.com/elastic/kibana-pull-request/builds/289496) succeeded 91da06e5cf07b14e73946ca03eead290ba109b7f
* :broken_heart: [Build #289364](https://buildkite.com/elastic/kibana-pull-request/builds/289364) failed 82d6b079b51540f0d3a0e74f26be513ea3b550a7
* :broken_heart: [Build #289348](https://buildkite.com/elastic/kibana-pull-request/builds/289348) failed 5dae5762b05b0b82e89aa810ce16973227c01dd3

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"running","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/300345","number":300345,"commit":"ee72370a5a87d28338702cfbe4f1506cb2ff982f"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/298027","number":298027,"commit":"48ded9315a027bc5a7eed2a1eb07741cb976af98"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/296996","number":296996,"commit":"f734bcd06dc7465d059f2a0fcf4e020e7335d258"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/289496","number":289496,"commit":"91da06e5cf07b14e73946ca03eead290ba109b7f"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/289364","number":289364,"commit":"82d6b079b51540f0d3a0e74f26be513ea3b550a7"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/289348","number":289348,"commit":"5dae5762b05b0b82e89aa810ce16973227c01dd3"}],"number":300345}
buildkite-pr-comment-->
